rfcs/20220404-troubleshooting-helpers/README.md

```diff
@@ -0,0 +1,304 @@
+However, this is trickier for the C API. In that case, if we break
+down a status into multiple statuses (e.g. instead of returning
+`invalid_argument`, we return `invalid_argument_X`), then it becomes
+hardly programmable if a user wants to catch all those.
+
+
+Another aspect is that it does not allow to help our user for
+non-failing diagnostics. For example, dispatching information cannot
+be obtained from status, as:
+- the primitive descriptor was created successfully, even though the
+  dispatched implementation is not the one expected by the user.
+- each implementation that was not dispatched might have a different
+  reason for not dispatching
+
+For the reasons above, we propose to leave the statuses as they
+currently are, so that they align with std::exception, and can be
+easily mapped from C to C++ API.
+
+
+## 2.2 Release mode verbose vs debug mode verbose
+Some debugging information might impact the performance of the
+library, in particular primitive creation and execution time.
+Currently, we already have such debugging facilities all over the
+place, and they are mostly controlled with dedicated macros, that are
+implementation specific (to name a few, we have `GEN_CONV_DEBUG`,
+`TR_DEBUG`, `DEBUGPRINT`).
+
+Here we propose to unify those macros. The main benefit is, for
+developers, that the mechanism to print and debug would be the same
+across implementations.
+
+This macro should be independent of `NDEBUG` as we might need to use
+this logging information on the release build. We propose to call this
+unified macro `ONEDNN_DEVEL_MODE`, and it should have a matching cmake
+option.
+
+We recommend that all messages related to internal logic go under
+`ONEDNN_DEVEL_MODE`. Only messages related to user visible APIs which
+are actionable by user should live outside `ONEDNN_DEVEL_MODE`.
+
+## 2.3 Verbose levels vs verbose flags
+
+Currently, we have some documented information messages we provide
+(primitive creation/execution time), and some undocumented
+(e.g. runtimes info and implementation details).  For the undocumented
+ones, we don't have alignment on which level they should print (some
+print at level 2 and above, other print at level 5 and above).
+
+The current verbose levels were designed with the assumption that each
+level of information was inclusive of the previous ones. If we want to
+increase the number of categories of information to print
+(e.g. dispatch info, perf, error info, debug info, ...), it will
+quickly bloat the verbose output, which can become an issue when using
+verbose for long running workloads.  More importantly, verbose
+messages might impact runtime (e.g. getting perf on GPU), so getting
+some information sometimes impacts runtime.
+
+We propose here to use flags in order to limit verbose bloat, and
+allow printing messages without impacting runtime when possible. This
+could take two forms:
+- a list of message types, like `ERROR,PROFILE,DISPATCH,DEVINFO`...
```

Comment:
I think it makes sense to unconditionally enable errors in case any of these types is enabled.

In general I don't think it's too bad to have inclusive levels (at least for the list of message types you mentioned), e.g. "dispatch" implying "profile" and "error" makes sense to me (the user requested more information so it's not a big deal to keep create/exec traces as well). This will make things a bit simpler.

Reply:
The main issue I have with inclusive-ness are:
1. output bloat. If a user wants dispatch info, those should apply only to primitive creation. If it includes primitive execution timings, it will quickly blow up in size.
2. Impact on execution flow: Currently, enabling profiling for GPU leads to serialization and host synchronizations. So ideally, we should not change the execution flow when we want to look at dispatch info, or print some jit ir dumps.
3. simplicity: When the number of info categories increases, if a user is interested by a couple of info only, they would have to know how they are ordered. It might be trivial to maintainers, but not to end-users.

Also, for those who don't care about bloat and impact on execution, I would expect they would just use `ALL` and be done with it (same as with level zero, most user rely on ZE_DEBUG=-1). When they send they log when filing an issue, it will just be more information for maintainers to understand what is going on.